### PR TITLE
AMDGPU: Verify function type matches when matching libcalls

### PR DESCRIPTION
**Pull request by arsenm**
Previously this would recognize a call to a mangled ldexp(float, float)
as a candidate to replace with the intrinsic. We need to verify the second
parameter is in fact an integer.

Fixes: SWDEV-501389

### CODE REVIEW

**Comment by arsenm** [comment]
* **#119043** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/119043?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/119043?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

**Review by changpeng — APPROVED** [review]
LGTM

File: llvm/lib/Target/AMDGPU/AMDGPULibFunc.cpp [hunk 1]
```diff
@@ -620,17 +620,17 @@ bool ItaniumParamParser::parseItaniumParam(StringRef& param,
...
-      .Case("ocl_image2darray" , AMDGPULibFunc::IMG2DA)
-      .Case("ocl_image1d"      , AMDGPULibFunc::IMG1D)
-      .Case("ocl_image2d"      , AMDGPULibFunc::IMG2D)
-      .Case("ocl_image3d"      , AMDGPULibFunc::IMG3D)
-      .Case("ocl_event"        , AMDGPULibFunc::DUMMY)
-      .Case("ocl_sampler"      , AMDGPULibFunc::DUMMY)
-      .Default(AMDGPULibFunc::DUMMY);
+    res.ArgType =
+        StringSwitch<AMDGPULibFunc::EType>(eatLengthPrefixedName(param))
+            .StartsWith("ocl_image1d_array", AMDGPULibFunc::IMG1DA)
```
Comment (ssahasra) [inline llvm/lib/Target/AMDGPU/AMDGPULibFunc.cpp:625]: Unrelated change in spellings? There's a function in this same file that produces the original spellings without the extra underscore.

**Review by ssahasra — CHANGES_REQUESTED** [review]

File: llvm/lib/Target/AMDGPU/AMDGPULibFunc.cpp (hunk 1 above)
Comment (arsenm) [reply llvm/lib/Target/AMDGPU/AMDGPULibFunc.cpp:625]: I can get away with the libraries not crashing but something is wrong with this. This should be the name of the type, which has an _ in the spec type names

Comment (ssahasra) [reply llvm/lib/Target/AMDGPU/AMDGPULibFunc.cpp:625]: Seems like this change should be split out into a separate patch that is complete by itself?

**Review by ssahasra — APPROVED** [review]

**Comment by arsenm** [comment]
### Merge activity

* **Dec 16, 12:57 AM EST**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/llvm/llvm-project/119043).

**Comment by jplehr** [comment]
I believe this patch turned the HIP buildbot red https://lab.llvm.org/buildbot/#/builders/123/builds/11310
So far, I was unable to reproduce the error locally.
It reaches an unreachable:
```
Unhandled param type
UNREACHABLE executed at /buildbot/llvm-project/llvm/lib/Target/AMDGPU/AMDGPULibFunc.cpp:961!
<...>
#10 0x000055d3fc651f98 llvm::AMDGPUMangledLibFunc::getFunctionType(llvm::Module const&) const (/buildbot/hip-vega20-0/clang-hip-vega20/llvm/bin/clang-19+0x130ff98)
#11 0x000055d3fc64f7c9 llvm::AMDGPULibFunc::isCompatibleSignature(llvm::Module const&, llvm::FunctionType const*) const (/buildbot/hip-vega20-0/clang-hip-vega20/llvm/bin/clang-19+0x130d7c9)
#12 0x000055d3fc649601 llvm::AMDGPULibCalls::fold(llvm::CallInst*) (/buildbot/hip-vega20-0/clang-hip-vega20/llvm/bin/clang-19+0x1307601)
#13 0x000055d3fc64a3df llvm::AMDGPUSimplifyLibCallsPass::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) (/buildbot/hip-vega20-0/clang-hip-vega20/llvm/bin/clang-19+0x13083df)
```

**Comment by jplehr** [comment]
> I believe this patch turned the HIP buildbot red https://lab.llvm.org/buildbot/#/builders/123/builds/11310 So far, I was unable to reproduce the error locally. It reaches an unreachable:
> 
> ```
> Unhandled param type
> UNREACHABLE executed at /buildbot/llvm-project/llvm/lib/Target/AMDGPU/AMDGPULibFunc.cpp:961!
> <...>
> #10 0x000055d3fc651f98 llvm::AMDGPUMangledLibFunc::getFunctionType(llvm::Module const&) const (/buildbot/hip-vega20-0/clang-hip-vega20/llvm/bin/clang-19+0x130ff98)
> #11 0x000055d3fc64f7c9 llvm::AMDGPULibFunc::isCompatibleSignature(llvm::Module const&, llvm::FunctionType const*) const (/buildbot/hip-vega20-0/clang-hip-vega20/llvm/bin/clang-19+0x130d7c9)
> #12 0x000055d3fc649601 llvm::AMDGPULibCalls::fold(llvm::CallInst*) (/buildbot/hip-vega20-0/clang-hip-vega20/llvm/bin/clang-19+0x1307601)
> #13 0x000055d3fc64a3df llvm::AMDGPUSimplifyLibCallsPass::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) (/buildbot/hip-vega20-0/clang-hip-vega20/llvm/bin/clang-19+0x13083df)
> ```

I checked the logs in more depth after not being able to reproduce the issue locally. May be a machine issue. Looking more into it.

**Comment by arsenm** [comment]
> I checked the logs in more depth after not being able to reproduce the issue locally. May be a machine issue. Looking more into it.

Any crash in the compiler cannot be a machine issue

#120068

**Comment by jplehr** [comment]
> Any crash in the compiler cannot be a machine issue

I still wonder why I cannot reproduce it though.

**Comment by arsenm** [comment]
> > Any crash in the compiler cannot be a machine issue
> 
> I still wonder why I cannot reproduce it though.

Release vs release with asserts build?

**Comment by jplehr** [comment]
> Release vs release with asserts build?

That could indeed be the case.
